### PR TITLE
Port opensearch operators

### PR DESCRIPTION
**Pull request by aljazerzen**
Ports `accept_opensearch` and `to_opensearch` operators

`accept_opensearch`:
- has previously been named `from_opensearch`,
- extracts a few HTTP-server related utilities into http.hpp,
- currently does not limit the number of concurrent requests

`to_opensearch`:
- very similar to `to_http`, uses same utilities.

Resolves TNZ-69
Resolves TNZ-70

### CODE REVIEW

File: libtenzir/builtins/operators/to_opensearch.cpp [hunk 1]
```diff
@@ -514,7 +495,334 @@ class opensearch_operator final : public crtp_operator<opensearch_operator> {
...
+    }
+    auto [ec, http_code] = req_->get<curl::easy::info::response_code>();
+    check(ec);
+    if (http_code < 200 or http_code > 299) {
+      diagnostic::warning("issue sending data. HTTP response code `{}`",
+                          http_code)
+        .note("response body: {}", response)
+        .primary(args_.operator_location)
+        .emit(dh);
+      lifecycle_ = Lifecycle::done;
```
[the remark this answers is not in the document]
Comment (aljazerzen) [reply libtenzir/builtins/operators/to_opensearch.cpp]: We should add proper retries. For now stopping the operator is ok.

File: libtenzir/builtins/operators/accept_opensearch.cpp [hunk 2]
```diff
@@ -0,0 +1,687 @@
...
+          (*response_signal)
+            ->send(Response{.status = 400,
+                            .content_type = std::string{bulk_content_type},
+                            .body = "{}"});
+          co_return;
+        }
+        if (args_.keep_actions) {
+          for (auto& slice : slices) {
+            if (slice.rows() > 0) {
+              co_await push(std::move(slice));
```
[the remark this answers is not in the document]
Comment (aljazerzen) [reply libtenzir/builtins/operators/accept_opensearch.cpp]: Well, that defeats the whole purpose of streaming responses. It's better to emit data multiple times instead of buffering the whole request in-memory.

File: libtenzir/src/tls_options.cpp [hunk 3]
```diff
@@ -75,6 +75,71 @@ auto parse_tls_version(std::string_view version) -> caf::expected<tls_version> {
...
+      return TLS1_1_VERSION;
+    case tls_version::v1_2:
+      return TLS1_2_VERSION;
+    case tls_version::v1_3:
+      return TLS1_3_VERSION;
+  }
+  TENZIR_UNREACHABLE();
+}
+
+struct resolved_tls_settings {
```
Comment (IyeOnline) [inline libtenzir/src/tls_options.cpp]: I dont really see the point of this struct. Its only used in one place and all its members are just a slightly shorter form of `opts.get_X(nullptr)`.

Personally I'd want to avoid having more code than we really need, but I also dont feel strongly about this.

File: libtenzir/builtins/operators/accept_http.cpp [hunk 4]
```diff
@@ -743,14 +541,51 @@ class AcceptHttp final : public Operator<void, table_slice> {
...
+          }
+        }
+        active_requests->clear();
+      }
+      co_await sleep_for(std::chrono::milliseconds{10});
+    }
+    if (server_) {
+      auto srv = std::move(*server_);
+      server_ = None{};
+      std::thread([srv = std::move(srv)]() {
```
Comment (IyeOnline) [inline libtenzir/builtins/operators/accept_http.cpp]: Having to push some non-constant destruction work off the executor has come up multiple times now.

I'm really not a fan of spawning a detatched thread like this, especially because to my understanding this "proper" shutdown handling wouldnt happen on e.g. a crash/assertion failure.

@jachris We briefly discussed this, but I dont think we decided on a good generic solution. Maybe we really should...

File: libtenzir/src/tls_options.cpp [hunk 5]
```diff
@@ -75,6 +75,71 @@ auto parse_tls_version(std::string_view version) -> caf::expected<tls_version> {
...
+  std::optional<located<std::string>> password;
+  std::optional<located<std::string>> ciphers;
+  bool skip_peer_verification = false;
+
+  auto private_key_file() const -> std::optional<located<std::string>> {
+    return keyfile ? keyfile : certfile;
+  }
+};
+
+auto resolve_tls_settings(tls_options const& opts, diagnostic_handler& dh)
```
Comment (IyeOnline) [inline libtenzir/src/tls_options.cpp]: All of of these accessor functions take a required argument for a reason: To access the node's configuration. If you dont provide this, the resolution will be wrong/ignore the config value.

The `nullptr` overload exists only for a select few cases to perform early validation.

You can get the config from `optctx.actor_system().config()`

File: libtenzir/src/http_server.cpp [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+                             diagnostic_handler& dh)
+  -> failure_or<wangle::SSLContextConfig> {
+  auto certfile = tls_opts.get_certfile(nullptr);
+  if (not certfile) {
+    diagnostic::error("`tls.certfile` is required when TLS is enabled")
+      .primary(primary)
+      .emit(dh);
+    return failure::promise();
+  }
+  auto keyfile = tls_opts.get_keyfile(nullptr);
```
Comment (IyeOnline) [inline libtenzir/src/http_server.cpp]: As mentioned elsewhere, you need to pass in the node config here.

File: libtenzir/include/tenzir/json_parser.hpp [hunk 7]
```diff
@@ -303,10 +305,54 @@ class ndjson_parser final : public parser_base {
...
+    if (split == 0) {
+      return {};
+    }
+    auto complete = blob{};
+    complete.insert(complete.end(), partial_.begin(), partial_.begin() + split);
+    auto trailing = blob{};
+    trailing.insert(trailing.end(), partial_.begin() + split, partial_.end());
+    partial_ = std::move(trailing);
+    auto parser = ndjson_parser{std::string{name}, dh, {}};
+    parser.parse_lines(chunk::make(std::move(complete)));
```
Comment (IyeOnline) [inline libtenzir/include/tenzir/json_parser.hpp]: I think it would be best if you did not extract this `parse_lines` from the old `from_opensearch`. Or rather, if we make the new code more efficient, there will be no use for common code anymore.

You are currently doing at least three copies of the decompressed message, where it would be possible to get away with zero in most cases.

As an idea:

* Make `http::decompress_playload` guarantee you that the buffer was over-sized to support the SIMDJSON padding requirement. You can do that by just over-allocating slightly. Since its in a cppfile in libtenzir, you can just include simdjson in there. Probably introduce a dedicated type for this though. Maybe just `struct simdjson_capable_blob : blob {};` does the job.
* Change the streaming_parser to work on this buffer type instead, removing any need for creating new, sized buffers (outside of the support for partial).
* Dont extract `ndjson_parser::parse_lines`, as its no longer needed in the streaming parser
* Leave `from_opensearch.cpp` unchanged.

**Review by IyeOnline — COMMENTED** [review]
Having once again spend too much time thinking about maybe-tangents, I'll leave a few intermediate comments before diving into the actual operators

File: libtenzir/src/tls_options.cpp (hunk 5 above)
Comment (aljazerzen) [reply libtenzir/src/tls_options.cpp]: I've seen that, but `accept_tcp` (old and new impl) only use nullptr. Should that also be fixed there?

And these changes are just refactoring of `tls_options::make_folly_ssl_context`, which uses `nullptr` everywhere.

File: libtenzir/builtins/operators/to_opensearch2.cpp [hunk 8]
```diff
@@ -0,0 +1,575 @@
...
+    }
+  }
+
+  ToOpenSearchArgs args_;
+  json_builder builder_;
+  Option<Box<HttpPool>> pool_ = None{};
+  std::string url_;
+  std::map<std::string, std::string> headers_;
+  MetricsCounter bytes_write_counter_;
+  Lifecycle lifecycle_ = Lifecycle::running;
```
Comment (IyeOnline) [inline libtenzir/builtins/operators/to_opensearch2.cpp]: I think you dont need this. Its only set to `done` on an error, right before returning from the respective coroutine.
In those cases emitting an error would however already cancel the operator, so no other member would be invoked.

File: libtenzir/builtins/operators/to_opensearch2.cpp [hunk 9]
```diff
@@ -0,0 +1,575 @@
...
+    return tenzir::detail::inspect_enum_str(f, x, {"running", "done"});
+  }
+
+  auto send_request(OpCtx& ctx) -> Task<void> {
+    TENZIR_ASSERT(pool_);
+    auto body = builder_.yield(ctx.dh());
+    if (body.empty()) {
+      co_return;
+    }
+    bytes_write_counter_.add(body.size());
```
Comment (IyeOnline) [inline libtenzir/builtins/operators/to_opensearch2.cpp]: For correctness, we should probably only update the metrics after the request went through.

File: libtenzir/builtins/operators/to_opensearch2.cpp [hunk 10]
```diff
@@ -0,0 +1,575 @@
...
+    if (result.is_err()) {
+      diagnostic::error("HTTP request failed: {}",
+                        std::move(result).unwrap_err())
+        .primary(args_.operator_location)
+        .emit(ctx);
+      lifecycle_ = Lifecycle::done;
+      co_return;
+    }
+    auto response = std::move(result).unwrap();
+    if (response.status_code < 200 or response.status_code > 299) {
```
Comment (IyeOnline) [inline libtenzir/builtins/operators/to_opensearch2.cpp:437]: I know this is the same behaviour as in the old executor, but I feel like we really should just be throwing away data with a mere warning here.

Personally I'd turn these transmission/remote failures into errors.

File: libtenzir/builtins/operators/to_opensearch2.cpp [hunk 11]
```diff
@@ -0,0 +1,575 @@
...
+  auto process(table_slice input, OpCtx& ctx) -> Task<void> override {
+    if (lifecycle_ != Lifecycle::running) {
+      co_return;
+    }
+    input = resolve_enumerations(std::move(input));
+    constexpr auto null_values
+      = []() -> generator<std::optional<std::string_view>> {
+      co_yield std::nullopt;
+    };
+    auto ids = resolve_str(
```
Comment (IyeOnline) [inline libtenzir/builtins/operators/to_opensearch2.cpp]: These should probably use `eval_as` instead of a handcrafted version

File: libtenzir/builtins/operators/to_opensearch2.cpp [hunk 12]
```diff
@@ -0,0 +1,575 @@
...
+          break;
+        }
+        case full: {
+          co_await send_request(ctx);
+          if (lifecycle_ != Lifecycle::running) {
+            co_return;
+          }
+          if (builder_.has_contents()) {
+            next_timeout_
+              = std::chrono::steady_clock::now() + args_.buffer_timeout.inner;
```
Comment (IyeOnline) [inline libtenzir/builtins/operators/to_opensearch2.cpp:351]: I think you also need a `notify_one()` here?

File: libtenzir/builtins/operators/accept_opensearch.cpp [hunk 13]
```diff
@@ -0,0 +1,754 @@
...
+            = http::make_decompressor(msg.content_encoding, failing_dh);
+          if (not decompressor) {
+            msg.response_signal->send(
+              Response{.status = 415,
+                       .content_type = std::string{bulk_content_type},
+                       .body = "{}"});
+            co_return;
+          }
+        }
+        auto requests = co_await active_requests_.lock();
```
Comment (IyeOnline) [inline libtenzir/builtins/operators/accept_opensearch.cpp]: I may be missing something here, but I dont think we need this lock?

Its only locked in `process_task` and `handle_forced_cancellation`, which afaict will never run concurrently.

File: libtenzir/builtins/operators/accept_opensearch.cpp [hunk 14]
```diff
@@ -0,0 +1,754 @@
...
+      [&](RequestBody msg) -> Task<void> {
+        auto response_signal = Option<Arc<ResponseSignal>>{None{}};
+        auto decompressor
+          = Option<std::shared_ptr<arrow::util::Decompressor>>{None{}};
+        auto is_action = true;
+        auto failed = false;
+        auto slices = std::vector<table_slice>{};
+        {
+          auto requests = co_await active_requests_.lock();
+          auto it = requests->find(msg.request_id);
```
Comment (IyeOnline) [inline libtenzir/builtins/operators/accept_opensearch.cpp]: Assuming that we even need the lock, we should not need to hold it after having `auto& req`. The backing standard container guarantees reference stability and we dont need to protect `req` itself, since we are guaranteed that only one `process_task` runs.

File: libtenzir/builtins/operators/accept_opensearch.cpp [hunk 15]
```diff
@@ -0,0 +1,754 @@
...
+          } else {
+            if (filtered.rows() > 0) {
+              co_await push(std::move(filtered));
+            }
+          }
+        }
+        auto requests = co_await active_requests_.lock();
+        auto it = requests->find(msg.request_id);
+        if (it != requests->end()) {
+          it->second.is_action = is_action;
```
Comment (IyeOnline) [inline libtenzir/builtins/operators/accept_opensearch.cpp]: This should also not be possible to fail, as nobody would have removed our entry from the requests map in the meantime, right?

File: libtenzir/builtins/operators/accept_opensearch.cpp [hunk 16]
```diff
@@ -0,0 +1,754 @@
...
+        }
+      });
+    co_await reader.read(
+      detail::narrow<uint32_t>(args_.get_max_request_size()));
+    if (started) {
+      co_await queue_->enqueue(RequestFinished{.request_id = request_id});
+    }
+    auto response = co_await response_signal->recv();
+    co_await folly::coro::co_reschedule_on_current_executor;
+    co_await queue_->enqueue(Noop{});
```
Comment (IyeOnline) [inline libtenzir/builtins/operators/accept_opensearch.cpp]: I dont understand this `Noop`.

File: libtenzir/builtins/operators/accept_opensearch.cpp [hunk 17]
```diff
@@ -0,0 +1,754 @@
...
+    co_await maybe_finish_draining();
+  }
+
+  auto handle_forced_cancellation() -> Task<void> {
+    if (server_) {
+      (*server_)->getServer().drain();
+    }
+    for (;;) {
+      while (auto message = message_queue_->try_dequeue()) {
+        if (auto* msg = std::get_if<RequestStarted>(&*message)) {
```
Comment (IyeOnline) [inline libtenzir/builtins/operators/accept_opensearch.cpp]:
```suggestion
        if (auto* msg = try_as<RequestStarted>(*message)) {
```

File: libtenzir/builtins/operators/accept_opensearch.cpp [hunk 18]
```diff
@@ -0,0 +1,754 @@
...
+          if (not msg->response_signal->has_sent()) {
+            msg->response_signal->send(
```
Comment (IyeOnline) [inline libtenzir/builtins/operators/accept_opensearch.cpp]: Do we want to respond with a 200 to requests that started when we are being canceled?

**Review by IyeOnline — COMMENTED** [review]
Went through the operators now. We should briefly talk about this (rather than go back and forth in comments), but it looks good.

File: libtenzir/builtins/operators/accept_opensearch.cpp (hunk 16 above)
Comment (aljazerzen) [reply libtenzir/builtins/operators/accept_opensearch.cpp]: It wakes the await_task, which wakes process_task, which invokes `maybe_finish_draining`.

File: libtenzir/builtins/operators/accept_http.cpp (hunk 4 above)
Comment (aljazerzen) [reply libtenzir/builtins/operators/accept_http.cpp]: No, this happens only when shutting down the operator via task cancellation. That is: when there is a `head` downstream or the pipeline is stopped manually.

Unfortunately, this path is very common in tests.

File: libtenzir/builtins/operators/accept_opensearch.cpp (hunk 13 above)
Comment (aljazerzen) [reply libtenzir/builtins/operators/accept_opensearch.cpp]: Very probably, you are right. When the operator stopping procedure was a bit different, it was needed.

File: libtenzir/builtins/operators/accept_opensearch.cpp (hunk 14 above)
Comment (aljazerzen) [reply libtenzir/builtins/operators/accept_opensearch.cpp]: This is needed to protect against a concurrent operator shutdown.

Comment (aljazerzen) [reply libtenzir/builtins/operators/accept_opensearch.cpp]: Again, not anymore, because `handle_forced_cancellation` cannot run concurrently with `process_task`.

File: libtenzir/builtins/operators/accept_opensearch.cpp (hunk 15 above)
Comment (aljazerzen) [reply libtenzir/builtins/operators/accept_opensearch.cpp]: Unless the request was cancelled and removed in the meantime.

File: libtenzir/builtins/operators/accept_opensearch.cpp (hunk 18 above)
Comment (aljazerzen) [reply libtenzir/builtins/operators/accept_opensearch.cpp]: No, it should be 503 here

File: test/fixtures/http_request.py [hunk 19]
```diff
@@ -34,7 +34,7 @@ class HttpRequestOptions:
     )
     tls: bool = False
     repeat: int = 1
-    expected_status: int = 200
+    expected_status: int | None = None
```
[the remark this answers is not in the document]
Comment (aljazerzen) [reply test/fixtures/http_request.py:37]: By default 503 is acceptable response to most requests, as it means the server is shutting down. This happens in tests, because `head 1` stops after first request and races against successful response.

Given that the clients should retry on 503 and we aim for "at-least-once" guarantee, this is acceptable.

File: libtenzir/builtins/operators/accept_opensearch.cpp [hunk 20]
```diff
@@ -0,0 +1,729 @@
...
+  bool failed = false;
+};
+
+class AcceptOpenSearch final : public Operator<void, table_slice> {
+public:
+  explicit AcceptOpenSearch(AcceptOpenSearchArgs args)
+    : args_{std::move(args)},
+      active_connections_{std::in_place, get_max_connections()} {
+  }
+
```
Comment (IyeOnline) [inline libtenzir/builtins/operators/accept_opensearch.cpp:314]: So we had a bit of a longer discussion about handling cancellation in a non-blocking but safe way. Obviously its non obvious and actually rather hard.

The current `handled_forced_cancellation` has a few issues. The `sleep` logic actually doesnt work, as the sleep will get canceled as well. Retrying indefinitely on some semaphore is (while I think correct here) incredibly scary. If the server gave you some awaitable telling you that it drained all active connections, it would be much prettier, but we also lack that.

We came to the realization that its probably OK if we just `forceStop()` the server. That will unceremoniously terminate the connection without a reply, but any client should be handle a connection/transport failure the same as a 503.

Importantly, its infinitely easier to implement:

```suggestion
  ~AcceptOpenSearch() noexcept {
    if (server_) {
      (*server_)->getServer().forceStop(); // technically also done by the destructor, but this is earlier than the execution of the spawned thread.
      std::thread([srv = std::exchange(server_,None{})] { 
      }).detach();
    }
  }
```

We also briefly discussed this as a generic feature, but for now this seems good enough.

Comment (aljazerzen) [reply libtenzir/builtins/operators/accept_opensearch.cpp:314]: Done

**Review by IyeOnline — APPROVED** [review]
